internal/webhook/external/server.go

```diff
@@ -71,3 +75,116 @@ func (s *server) Serve(ctx context.Context, l net.Listener) error {
 		return err
 	}
 }
+
+func (s *server) refreshWarehouseHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := (r.Context())
+	logger := logging.LoggerFromContext(ctx)
+	logger.Debug("refresh warehouse handler called", "path", r.URL.Path)
+	var projectConfigs kargoapi.ProjectConfigList
+	err := s.client.List(
+		ctx,
+		&projectConfigs,
+		client.MatchingFields{
+			indexer.ProjectConfigsByWebhookReceiverPathsField: r.URL.Path,
+		},
+	)
+	if err != nil {
+		logger.Error(err, "failed to list project configs")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if len(projectConfigs.Items) == 0 {
+		logger.Info("no project configs found")
+		http.Error(w, "no project configs found for the request", http.StatusNotFound)
+		return
+	}
+
+	// There will always be only one project config
+	// because two project configs can never have the same
+	// receiver path because the receiver path is
+	// a hash of the project name, provider and secret.
+	pc := projectConfigs.Items[0]
+
+	var wrc *kargoapi.WebhookReceiverConfig
+	if len(pc.Spec.WebhookReceivers) == 1 {
+		wrc = &pc.Spec.WebhookReceivers[0]
+	} else {
+		wrc, err = s.getWebhookReceiverConfig(r.URL.Path, pc)
+		if err != nil {
+			logger.Error(err, "failed to find webhook receiver config")
+			http.Error(w, err.Error(), http.StatusNotFound)
+			return
+		}
+	}
+
+	logger.Debug("webhook receiver config found", "webhook-receiver-config", wrc)
+	switch {
+	case wrc.GitHub != nil:
+		var secret corev1.Secret
+		err := s.client.Get(ctx,
+			client.ObjectKey{
+				Name:      wrc.GitHub.SecretRef.Name,
+				Namespace: pc.Namespace,
+			},
+			&secret,
+		)
+		if err != nil {
+			logger.Error(err, "failed to get github secret")
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("failed to get github secret %q: %w", wrc.GitHub.SecretRef.Name, err),
+					http.StatusNotFound,
+				),
+			)
+			return
+		}
+		token, ok := secret.Data["token"]
+		if !ok {
+			logger.Error(
+				errors.New("failed to get github token from secret"),
+				"no value for 'token' key",
+			)
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					errors.New("missing github token in secret"),
+					http.StatusInternalServerError,
+				),
+			)
+			return
+		}
+		githubHandler(s.client, string(token))(w, r)
+	default:
+		http.Error(w, "not found", http.StatusNotFound)
+	}
+}
+
+func (s *server) getWebhookReceiverConfig(
```

Comment:
I think there's a more efficient approach to this method. As written, you are re-generating the path every time to compare it to the path that was provided as an argument. The hashing, not to mention the underlying Secret retrieval necessary to do that, are more expensive than the alternative...

Loop through the statuses to match the path provided as an argument to a webhook receiver, and then you will know the index of the webhook receiver config that you want from the spec. It gets even better after [this](https://github.com/akuity/kargo/pull/4000/files#r2101075827) is addressed because then you can dispense with the assumption that the entries in the two lists are ordered the same.